packages/nx/src/utils/package-manager.ts

```diff
@@ -157,32 +181,26 @@ export function resolvePackageVersionUsingRegistry(
  * installing it in a temporary directory and fetching the version from the
  * package.json.
  */
-export function resolvePackageVersionUsingInstallation(
+export async function resolvePackageVersionUsingInstallation(
   packageName: string,
   version: string
-): string {
-  const dir = dirSync().name;
-  const npmrc = checkForNPMRC();
-
-  writeJsonFile(`${dir}/package.json`, {});
-  if (npmrc) {
-    // Copy npmrc if it exists, so that npm still follows it.
-    copyFileSync(npmrc, `${dir}/.npmrc`);
-  }
-
-  const pmc = getPackageManagerCommand();
-  execSync(`${pmc.add} ${packageName}@${version}`, { stdio: [], cwd: dir });
-
-  const packageJsonPath = require.resolve(`${packageName}/package.json`, {
-    paths: [dir],
-  });
-  const { version: resolvedVersion } = readJsonFile(packageJsonPath);
+): Promise<string> {
+  const dir = createTempNpmDirectory();
 
   try {
-    unlinkSync(dir);
-  } catch {
-    // It's okay if this fails, the OS will clean it up eventually
+    const pmc = getPackageManagerCommand();
+    await execAsync(`${pmc.add} ${packageName}@${version}`, { cwd: dir });
+
+    const packageJsonPath = require.resolve(`${packageName}/package.json`, {
+      paths: [dir],
+    });
+
+    return readJsonFile<PackageJson>(packageJsonPath).version;
```

Comment:
Ohhh shit I forgot about this, I remember having a talk about this months ago with friends. Great damn catch.

Reply:
What do you think about abstracting this behavior with something like:

```ts
async function withTempNpmDirectory<T>(action: (dir: string) => Promise<T>): Promise<T> {
  const dir = createTempNpmDirectory();

  let result: T;

  try {
    result = await action(dir);
  } finally {
    try {
      await remove(dir);
    } catch {
      // It's okay if this fails, the OS will clean it up eventually
    }
  }

  return result;
}
```

Is this overkill or fine?

Reply:
It looks like I'm actually wrong? The code definitely reads weird, so I'd still recommend moving it. 

According to https://eslint.org/docs/rules/no-unsafe-finally, the return statement in `try` would be suspended until after the `finally` block ran. This is unintuitive, and I'd consider it a JS quirk.

Reply:
I wouldn't put it in the with block like that, just because it doesn't line up with a lot of the other code in the repo. Its not bad, but may be too clever. If anything, I'd have the createTmp method return 
```
{
  "path": "/tmp/some-folder",
  "cleanup": () => {}
}
```

Reply:
Ok so I'll revert the changes I've made 😅 the problem is that the calling code of `createTempNpmDirectory` needs to handle the cleanup itself, not what the clean is, so returning a cleanup function is good, but it does not fix the problem that bothers me in the first place (the need to call cleanup manually).